### PR TITLE
Add variant for old x86_64 CPUs without SSSE 4.2 or AVX instructions

### PR DESCRIPTION
**Pull request by pdecat**
Resolves https://github.com/blakeblackshear/frigate/issues/49

Dynamically build tensorflow tflite wheel on target platform and use official blakeblackshear/frigate:0.8.1 as base image.

### CODE REVIEW

**Comment by pdecat** [comment]
Removed draft status as it was reported as working: https://github.com/blakeblackshear/frigate/issues/49#issuecomment-771867096

**Comment by jonty-comp** [comment]
This is just what I needed - looking to use an old HP Microserver as an NVR and the Turion processor doens't support AVX.  Since this PR is for the HassOS integration, will there need to be a separate change for those of us who just want to run the Frigate instance on its own in a Docker container?

**Comment by pdecat** [comment]
> Since this PR is for the HassOS integration, will there need to be a separate change for those of us who just want to run the Frigate instance on its own in a Docker container?

You should be able to download the Dockerfile locally and `docker build` it.

**Comment by pdecat** [comment]
This does not work as expected when a Coral TPU is used: https://github.com/blakeblackshear/frigate/issues/695#issuecomment-785463546

More work needed to actually rebuild the `libedgetpu.so.1.0` library...

**Comment by pdecat** [comment]
This has been confirmed to work by two independent users: see https://github.com/blakeblackshear/frigate/issues/695#issuecomment-787892426 & https://github.com/blakeblackshear/frigate/issues/695#issuecomment-788494430

**Comment by pdecat** [comment]
Hi @blakeblackshear, would you be interested in merging this variant into you main repository?

I think the only change needed before merging, would be to rename `frigate_beta_old` to `frigate_oldcpu` or something like that given `frigate_beta` is obsolete. WDYT?

**Comment by blakeblackshear** [comment]
I plan to merge this in after I have a chance to review it more closely.

**Comment by LeoI5** [comment]
> I plan to merge this in after I have a chance to review it more closely.

I've checked it on Mac Mini 2007 (MacMini2,1). Addon works fine (no Coral connected).
I still wait Google Coral from seller (huge queue). 
There was some problem with installation. I tried to install some times. Process was in progress without finish.
After the third attempt addon has been installed and start to work. Probably my Mac is too slow to install this addon on the first try.

**Comment by albertcaba** [comment]
What is the status of this? Why is this not merged?

**Comment by albertcaba** [comment]
Hello, how can I use this? I tried adding https://github.com/pdecat/frigate-hass-addons to the addon store in Hassio, I could add it, but when I try to install the old cpu variant, the installation never ends. Also the link Frigate NVR Beta for old x86_64 CPUs is dead. Do I need to create my own docker container? Or am I doing something wrong? Thank you in advance!

**Comment by pdecat** [comment]
> when I try to install the old cpu variant, the installation never ends

During installation, a container image is locally built, and the tensorflow dependencies are recompiled with your own CPU.
Compilation can take a long time depending on the performance of your system. Something like 30 minutes to complete maybe.

If it fails, you should see the error messages in the supervisor logs with `docker logs hassio_supervisor`.

**Comment by albertcaba** [comment]
Hi Patrick,
Thank you for your help, now it is working, sorry for bothering you with a silly mistake! Thank you for your work! :)

**Comment by ManuelPerrot** [comment]
This branch has solved my problem for running it on an old laptop with coral pci card.
git clone & docker build locally ^^
Thanks a lot 👍 

Is it possible to add this to the supported architectures and push updated images to docker hub going forward?

**Comment by chherman0** [comment]
I have an old AMD Phenom 2 in my little home server and keep getting the fatal error. Not quite sure how to make this thing work. I am really new to the docker world, only have a few running right now, and have no idea how to get the older cpu docker running. I am using OMV as my host OS and no coral yet. Any guides on how to get the older cpu version up and running?

File: frigate_oldcpu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+RUN cd tensorflow && git checkout 8ae2eecf4f70b5efb55a108f2b0000ea6cad3e05
+RUN pip3 install -U \
+    # pin to the same version that frigate pins, see https://github.com/blakeblackshear/frigate/blob/v0.8.1/docker/Dockerfile.wheels#L27
+    numpy==1.19.4 \
+    pybind11
+RUN bash tensorflow/tensorflow/lite/tools/make/download_dependencies.sh
+RUN bash tensorflow/tensorflow/lite/tools/pip_package/build_pip_package.sh
+
+FROM blakeblackshear/frigate:0.8.4-${BUILD_ARCH}
+COPY --from=build_libedgetpu /libedgetpu1-max_14.0_amd64.deb /
```
Comment (pdecat) [inline frigate_oldcpu/Dockerfile:42]: TODO: update to v15 (or even v16) as debian packaging files for newer releases are now available.

See https://github.com/google-coral/libedgetpu/issues/14#issue-819832572

Comment (chherman0) [reply frigate_oldcpu/Dockerfile:42]: Unfortunately, I am not sure what to do with that. Ill keep looking and researching. Thank you

Comment (pdecat) [reply frigate_oldcpu/Dockerfile:42]: This comment was not addressed to you, just a reminder for myself.

**Comment by ManuelPerrot** [comment]
> I have an old AMD Phenom 2 in my little home server and keep getting the fatal error. Not quite sure how to make this thing work. I am really new to the docker world, only have a few running right now, and have no idea how to get the older cpu docker running. I am using OMV as my host OS and no coral yet. Any guides on how to get the older cpu version up and running?

Hi,
First you need to git clone this PR as the code is not merged to the main branch yet.
This needs to be done on the system you are planning to run the docker container:
`git clone https://github.com/pdecat/frigate-hass-addons/tree/oldcpu`

Then create the docker image for your system by running something like:
`sudo docker build -t frigate-no-avx:8.4 -f /tmp/frigate-hass-addons/frigate_beta_oldcpu/Dockerfile /tmp/`
It will create a local image named "frigate-no-avx:8.4" (might be few errors and warning along the way and take some time)

Then you can reference it in your docker-compose to deploy it:
`image: frigate-no-avx:8.4`

**Comment by chherman0** [comment]
Thank you for the reply. When running the first command i get a 

root@openmediavault:~# git clone https://github.com/pdecat/frigate-hass-addons/tree/oldcpu
Cloning into 'oldcpu'...
fatal: repository 'https://github.com/pdecat/frigate-hass-addons/tree/oldcpu/' not found

then i try this one and get this response.

root@openmediavault:~# git clone https://github.com/pdecat/frigate-hass-addons/tree/oldcpu/frigate_oldcpu
Cloning into 'frigate_oldcpu'...
fatal: repository 'https://github.com/pdecat/frigate-hass-addons/tree/oldcpu/frigate_oldcpu/' not found

**Comment by dannysauer** [comment]
For whoever finds this later and wants some simple directions which work with the current state of things. ;)
```
git clone https://github.com/pdecat/frigate-hass-addons
cd frigate-hass-addons/       
git checkout oldcpu
docker build -t frigate-no-avx:8.4 frigate_oldcpu
```
...then go make yourself a small meal, because someone who needs this will have plenty of time to kill. ;)

**Comment by chherman0** [comment]
@dannysauer, thanks! Ill try it out

**Comment by pdecat** [comment]
I've reproduced the issue and pushed a fix that should resolve it: https://github.com/blakeblackshear/frigate-hass-addons/pull/12/commits/e683d5d63c56ab8be863b0a4d3849173d1c18866

Testing on old CPU platform in progress...

**Comment by pdecat** [comment]
I can confirm this is now working on my old CPU platform.

However, with 4GB of RAM, I had to increase my swap to 32 GB for the build to succeed without out of memory errors (it failed with 16 GB of swap).

**Comment by ozrouter** [comment]
I couldn't get the latest version to build on a old noavx pc.
So I built it on a new PC with avx, no edgetpu and cpu detector worked fine. (Proxmox LXC container)
Then migrated this container to the old no avx PC and it ran fine with the CPU detector. 

Changed to the edgetpu and it errors. (I have the old version running fine with edgetpu on this PC)
**When it originally builds must it have the edgetpu installed to work with it?**

> [2021-11-04 10:20:09] detector.coral                 INFO    : Starting detection process: 217
Fatal Python error: Illegal instruction
Thread 0x00007f1409ce7700 (most recent call first):
  File "/usr/lib/python3.8/multiprocessing/connection.py", line 368 in _send
  File "/usr/lib/python3.8/multiprocessing/connection.py", line 411 in _send_bytes
  File "/usr/lib/python3.8/multiprocessing/connection.py", line 200 in send_bytes
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 245 in _feed
  File "/usr/lib/python3.8/threading.py", line 870 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap
Current thread 0x00007f141a7ff740 (most recent call first):
  File "/usr/lib/python3.8/ctypes/__init__.py", line 373 in __init__
  File "/usr/lib/python3.8/ctypes/__init__.py", line 451 in LoadLibrary
  File "/usr/local/lib/python3.8/dist-packages/tflite_runtime/interpreter.py", line 89 in __init__
  File "/usr/local/lib/python3.8/dist-packages/tflite_runtime/interpreter.py", line 160 in load_delegate
  File "/opt/frigate/frigate/edgetpu.py", line 64 in __init__
  File "/opt/frigate/frigate/edgetpu.py", line 156 in run_detector
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108 in run
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315 in _bootstrap
  File "/usr/lib/python3.8/multiprocessing/popen_fork.py", line 75 in _launch
  File "/usr/lib/python3.8/multiprocessing/popen_fork.py", line 19 in __init__
  File "/usr/lib/python3.8/multiprocessing/context.py", line 277 in _Popen
  File "/usr/lib/python3.8/multiprocessing/context.py", line 224 in _Popen
  File "/usr/lib/python3.8/multiprocessing/process.py", line 121 in start
  File "/opt/frigate/frigate/edgetpu.py", line 241 in start_or_restart
  File "/opt/frigate/frigate/edgetpu.py", line 210 in __init__
  File "/opt/frigate/frigate/app.py", line 187 in start_detectors
  File "/opt/frigate/frigate/app.py", line 326 in start
  File "/opt/frigate/frigate/__main__.py", line 17 in <module>
  File "/usr/lib/python3.8/runpy.py", line 87 in _run_code
  File "/usr/lib/python3.8/runpy.py", line 194 in _run_module_as_main

**Comment by ozrouter** [comment]
I created the swap but it still crashed. It didn't appear to use the swap though. Is there a way to force it perhaps?

**Comment by pdecat** [comment]
> I created the swap but it still crashed. It didn't appear to use the swap though. Is there a way to force it perhaps?

When you say it crashed, did you get out of memory errors in kernel message logs with `dmesg`?

You can check whether the swap is active and used using the `free` command.

**Comment by ozrouter** [comment]
I have seen the memory errors **without** the extra swap.
With 32gb swap and 4gb ram the build ended up crashing and needing a full reboot after 10 mins.

This latest commit **works with no avx CPU detector** (compiled from another pc).
so I guess the edgetpu section of the docker file is failing in this case.

is there a method of logging the failing part of the docker file on the docker add on build process?

**Comment by pdecat** [comment]
> This latest commit works with no avx CPU detector (compiled from another pc).

Your other PC may have other CPU instructions that are not supported by your target system.

**Comment by pdecat** [comment]
FYI, I ordered a second Dual Edge TPU which should be delivered in the coming weeks. I'll resume work on this PR once I install it in my test system.

**Comment by pdecat** [comment]
> is there a method of logging the failing part of the docker file on the docker add on build process?

@ozrouter you should see the build errors in the supervisor logs:

```
docker logs hassio_supervisor  -f
```

**Comment by pdecat** [comment]
Hi, I've finally completed the migration of my Home Assistant installation from a VM to a physical machine with an old CPU and can confirm the docker image built from this branch is working fine.

Some details about my system:
```
patrick@ha:~$ grep '\(model name\|flags\)' /proc/cpuinfo
model name      : Intel(R) Core(TM)2 Duo CPU     E4400  @ 2.00GHz
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ht tm pbe syscall nx lm constant_tsc arch_perfmon pebs bts rep_good nopl cpuid aperfmperf pni dtes64 monitor ds_cpl est tm2 ssse3 cx16 xtpr pdcm lahf_lm pti dtherm
model name      : Intel(R) Core(TM)2 Duo CPU     E4400  @ 2.00GHz
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ht tm pbe syscall nx lm constant_tsc arch_perfmon pebs bts rep_good nopl cpuid aperfmperf pni dtes64 monitor ds_cpl est tm2 ssse3 cx16 xtpr pdcm lahf_lm pti dtherm
```

```
patrick@ha:~$ free -m
               total        used        free      shared  buff/cache   available
Mem:            3863        2279         126         131        1458        1277
Swap:          32767         361       32406
```

```
[services.d] starting services
[services.d] done.
[2021-11-22 01:25:19] frigate.app                    INFO    : Starting Frigate (0.9.4-26ae608)
[2021-11-22 01:25:19] frigate.app                    INFO    : Creating directory: /tmp/cache
Starting migrations
[2021-11-22 01:25:19] peewee_migrate                 INFO    : Starting migrations
There is nothing to migrate
[2021-11-22 01:25:19] peewee_migrate                 INFO    : There is nothing to migrate
[2021-11-22 01:25:19] frigate.mqtt                   INFO    : MQTT connected
[2021-11-22 01:25:19] detector.coral0                INFO    : Starting detection process: 213
[2021-11-22 01:25:19] frigate.edgetpu                INFO    : Attempting to load TPU as pci:0
[2021-11-22 01:25:19] detector.coral1                INFO    : Starting detection process: 215
[2021-11-22 01:25:19] frigate.app                    INFO    : Output process started: 216
[2021-11-22 01:25:19] frigate.edgetpu                INFO    : Attempting to load TPU as pci:1
[2021-11-22 01:25:19] ws4py                          INFO    : Using epoll
[2021-11-22 01:25:19] frigate.app                    INFO    : Camera processor started for camera0: 223
[2021-11-22 01:25:19] frigate.app                    INFO    : Camera processor started for camera1: 225
[2021-11-22 01:25:19] frigate.app                    INFO    : Camera processor started for camera2: 228
[2021-11-22 01:25:19] frigate.app                    INFO    : Capture process started for camera0: 232
[2021-11-22 01:25:19] frigate.app                    INFO    : Capture process started for camera1: 236
[2021-11-22 01:25:19] frigate.app                    INFO    : Capture process started for camera2: 244
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
[2021-11-22 01:25:19] frigate.edgetpu                INFO    : TPU found
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
[2021-11-22 01:25:19] frigate.edgetpu                INFO    : TPU found
[2021-11-22 01:25:20] ws4py                          INFO    : Using epoll
```

image

**Comment by pdecat** [comment]
I've rebased this PR on main to update frigate to 0.10.0.

Also, I've made a change to limit the number of jobs during build to avoid hogging the CPU of the target host.

**Comment by ManuelPerrot** [comment]
Thanks! That's still working for me 👍

**Comment by off112a** [comment]
Thanks @pdecat This is working perfect for me :+1: Do you have any plans to update this to 0.11?

**Comment by pdecat** [comment]
Rebased on current main to update frigate to 0.11.1.

PS: I'm no longer using this as I upgraded to a new server, so I did not actually test it.

**Comment by off112a** [comment]
Thank you. It's working 🙂

**Comment by elperepat** [comment]
Any chance this might be updated to 0.12?

**Comment by felipecrs** [comment]
What are these old CPUs that you are using?

**Comment by elperepat** [comment]
Xeon X5365

**Comment by felipecrs** [comment]
Got it. Thank you. That's old indeed.

However, as I mentioned, I use a J4125 CPU which also does not support AVX instructions according to ARK and never had any problems using regular Frigate builds.

Maybe the issue is not about lack of AVX after all, but something else.

**Comment by pdecat** [comment]
Rebased on current main to update frigate to 0.12.0.

PS: I'm no longer using this as I upgraded to a new server, so I did not actually test it.

**Comment by blastrock** [comment]
I don't have a supervised installation of home assistant, but I copy pasted your Dockerfile and built it myself, it worked on my Athlon II X2, thanks!

**Comment by kepten** [comment]
I use this variant on two mini PCs for more than a year now without a problem.
@blakeblackshear is there any reason not to merge it?

**Comment by blakeblackshear** [comment]
It is out of date at this point, and I don't have any way to test/support it.

**Comment by kepten** [comment]
@blakeblackshear If by out of date, you mean the older Frigate version, I have fixed it, [here](https://github.com/kepten/frigate-hass-addons) is my version and the PR I just opened: https://github.com/pdecat/frigate-hass-addons/pull/2

**Comment by pdecat** [comment]
:tada: 

Thanks @blakeblackshear and all :)

**Pull request merged**